Clean up module hierarchy and add fuzz tests for Laws.

erlandsona opened the pull request.

erlandsona committed:
Clean up module hierarchy and add fuzz tests for Laws.

erlandsona committed:
Now all the documentation is passing :thumbsup:

erlandsona commented:
[For an example of the new `name` function see here](https://github.com/erlandsona/elm-ui-form/blob/main/src/Pages/Home_.elm#L126)

Basically using a lens as a unique identifier to a key in a dictionary gives a couple things:
1. roughly the same semantics of [choonkeat/formdata](https://package.elm-lang.org/packages/choonkeat/formdata/latest/) without the callers having to define a custom Sum type representing the keys in the record coming out of the form... 

AND

2. Fixes https://github.com/etaque/elm-form/issues/97 by using the phantom type to ensure that only lenses that point to keys in the output record can be used to represent data in the underlying `Dict String String`

erlandsona committed:
Add isProperty fuzzes for Array actions...

bChiquet reviewed (commented):
There seems to be lots of changes in here, and it's going to be hard for me to process everything in one go.

Here are the themes that I can see in the PR:
* add accessors for list, tuples and Arrays
* modify makeOneToOne and makeOneToN to add a description field in them
* restructure code comments
* restructure the file hierarchy (i.e. put everything in one file)
* add tests
* remove the lazy accessor operations

Some of those are great, like the new accessors for data structures and tests
Some have costly tradeoffs, like the modifications to addOneToX or the namechange for the list accessor.
Some seem like oversights, like the removal of the recently added lazy operations.

Since it is hard to handle all of these things in a single PR acceptance or rejection, would you agree to work together in order to break this big change in separate small PRs?

The PRs that would be easiest to pass first would be the tests and the new accessors.

erlandsona committed:
Make description field backwards compat with original libary design.

erlandsona commented:
- [x] add accessors for list, tuples and Arrays
- [x] modify makeOneToOne and makeOneToN to add a description field in them
   - I've made this backwards compatible by defaulting to `""` and exposing `makeOneToN_` and `makeOneToOne_` resp.
- restructure code comments
   - Really the only change is `-- return` to `-->` to make the comments parsable by elm-verify-example which is what generated the `VerifyExample/` directory with tests for all the the doc comments. The comments themselves shouldn't have changed much and I made an effort to hold to the style of the comments in the docs for the new accessors I added.
- restructure the file hierarchy (i.e. put everything in one file)
   - This cleans up the api by making the common accessors available from the same import.
    ```elm
    import Accessors exposing (Relation, get, set, over)
    import Accessors.Library exposing (try) -- just felt sort of noisy 🤷‍♂️ 
    ```
    vs
    ```elm
    import Accessors exposing (Relation, get, set, over, try)
    ```
- add tests
   - Yup, after watching most of chiroptical's streaming through Chris Penners Optics By Example implementing Fuzz tests for the laws seemed like low hanging fruit that could help me better understand the distinctions between the different types of accessors. But most of the File additions came from elm-verify-examples
   - I pulled inspiration for the specs from [Lens Properties](https://hackage.haskell.org/package/lens-properties-4.11.1/docs/src/Control-Lens-Properties.html#isLens) of which I couldn't seem to implement Iso, Prism, Traversal cause Traversal is just `over` and I still haven't really figured out how elm-accessors represents `Prism`, `Iso` also seems weird and probably unnecessary in an Elm setting.
- remove the lazy accessor operations
  - So I removed the module but I merged the implementation since the Lazy version seemed like it was just doing memoization. 
  - That said, I'm not sure what the implications of the lazy version are yet... and if we actually need both, I'd probably do the same thing I did with `makeOneToOne_`... I'm not a huge fan of underscored methods but the Builder pattern seemed overkill 🤷‍♂️

erlandsona committed:
Include a Lens alias for defining Lenses over extensible records in Codegen.

erlandsona committed:
Messin' :man_shrugging:

erlandsona committed:
Rename test Lenses and workin' on more general types for get/set/over so they unify and work better for building higher-order lenses.

erlandsona committed:
Continuing to figure out the proper generalizations of the get/set/over signatures so they can work with more stuff...

erlandsona committed:
wip: proper general type aliases.

bChiquet commented:
> I've made this backwards compatible by defaulting to "" and exposing makeOneToN_ and makeOneToOne_ resp.

Thanks for that.
I'm still looking into the use case in your module.
Sorry if this causes a bit of delay for merging this feature.

> restructure code comments

There are some nitpicks I'd like to make on comments/doc, but since there are other, more structural issues, I will probably come back on it later as part of separate topics.

> restructure the file hierarchy (i.e. put everything in one file)
> This cleans up the api by making the common accessors available from the same import.

While I understand your motivation on this, moving functions to a separate module breaks the API for people that already use it. Whenever possible, I would prefer to push new versions that don't require any work for existing users, especially for aesthetic changes that they may disagree with.

For people that prefer it, this design can also be achieved locally using a single `MyProject.Accessors` module that reexposes the library in a way that suits them.

> add tests

I'm in the process of working out the tests: I'm new to `elm-verify-examples` so I'm looking how it works.

> remove the lazy accessor operations 

I'm not sure about the impacts of moving to the lazy version either, especially in terms of performance for people using accessors on large data structures. That is why I would prefer to keep both for now, and merge versions after some time experimenting with them.

erlandsona committed:
Add Library for Backwards compat.

erlandsona committed:
Add back Lazy versions.

erlandsona commented:
> Sorry if this causes a bit of delay for merging this feature.
I'm in no rush, I've got this copy/pasted locally so merging this isn't a blocker for us at least, just trying to do my part to participate in open source and contribute back to upstream.
> some nitpicks I'd like to make on comments/doc
🤷 sure.
> Module Heirarchy
I've updated this to make it backwards compatible as well. 
> Accessors.Lazy 
Just fixed this as well. Strict is default again as it used to be and the Lazy module is back, based on the new generalized strict versions.

FWIW, the fact that packages in Elm have enforced Semver means that for organizations that don't agree with changes in library design they can either stay on their existing version or fork / release their own versions. I'm also totally open to releasing these changes under my own namespace but I like to avoid splitting hairs when I can.

For now I've made all the changes backwards compatible. Or at least they should be. And hopefully that calms any unease about the additions. If you go to publish and it warns about breaking changes feel free to let me know so I can come up with more ideas on how to keep it backwards compatible but personally I'm not really bothered upgrading major versions in Elm cause the compilers so helpful it's way less of a burden than it would be in other languages.

erlandsona commented:
Also, I'm still not settled on the `type alias Setable`, `type alias Getable`, `type alias Lens_`, etc... so would love your ideas on how to improve on those.
The goal being that it's non-trivial to work out the compiler errors when certain compositions don't compose correctly because of the laws and the compiler is forcing the developer to read function signatures for what look like dot-properties at the call site. 
Just attempting to make things easier to reason about.

erlandsona commented on tests/SpecLazy.elm:
```diff
@@ -1,188 +0,0 @@
-module SpecLazy exposing (suite)
```
This seemed like it was just copy/pasted over from `tests/Spec` which is why I deleted it... but might be useful since Lazy is back? If this is just testing to verify the Lazy versions still work fine, what I'm curious about is if we can't actually test the laziness of the versions in the Lazy module more specifically somehow?

erlandsona committed:
Tests for eachIdx

erlandsona committed:
Adds a few more combinators and associated docs & doc tests.

erlandsona commented:
👋 pushed up a few more combinators.

Was hoping you might weigh in on the idea of defaulting as it pertains to "Structure Preserving" vs "Non Structure Preserving" combinators. EG:
```elm
[1,2,3]
    |> set (at 1) 77 -- `at` here is structure preserving so you can't use `set (at 1) Nothing` to "delete" a key as that would make this an "unlawful"(?) lens?
    == [1,77,3]
```
Note that because `at` is structure preserving you have to use `or` for "defaulting"
```elm
[1,2,3]
    |> get (at 10 |> or 9999)
    == 9999
```
VS: Dict & key & def
```elm
Dict.fromList [("foo", 1), ("bar", 2), ("baz", 3)]
    |> set (key "new") (Just 4) -- is NON structure preserving
    |> set (key "foo") Nothing -- can be used instead of `Dict.remove`
```
But defaulting for `key` doesn't work with `or` so we use `def` instead.
```elm
Dict.fromList [("foo", 1), ("bar", 2), ("baz", 3)]
    |> get (key "BAD" << def 77) -- `def` here only works composed with NON structure preserving lenses...
    == 77
```

bChiquet commented:
As per our Slack discussion, this PR will be hard to merge, so focusing on the questions for a potential fork:

> Was hoping you might weigh in on the idea of defaulting as it pertains to "Structure Preserving" vs "Non Structure Preserving" combinators.

In your example, defaulting for getters can be done without a lens: 
```
[1,2,3]
    |> get (at 10) |> Maybe.withDefault 9999
    == 9999
```
If you take more complex examples: 
```
Just [1,2,3]
    |> get (try << at 10) |> Maybe.withDefault 9999
    == compilation error
```
In this situation, classic defaulting is not performing what we want, so your idea of a default value may make sense. It is not obvious to me, however, how your example would work. The case, which you cite later on, to make a `withDefault` lens is [more promising](https://ellie-app.com/hmg5952KJfZa1).

For setters, you don't have such an issue since the `wrap` is never used. In my opinion you would be better off using existing options: 
```
Dict.fromList [("foo", 1), ("bar", 2), ("baz", 3)]
    |> set (key "new") (Just 4) -- is NON structure preserving
```
If the idea is to take a `Maybe a` instead of an `a` as input for the lens, that makes it pretty confusing to know what you actually insert from reading the code.
```    
Dict.fromList [("foo", 1), ("bar", 2), ("baz", 3)]
    |> set (key "foo") Nothing -- can be used instead of `Dict.remove`
```
It is not obvious to me what you gain compared to: 
```
Dict.fromList [("foo", 1), ("bar", 2), ("baz", 3)]
    |> Dict.remove "foo"
```
or, in the case of more nested stuff:

```
Just (Dict.fromList [("foo", 1), ("bar", 2), ("baz", 3)])
    |> over (try) (Dict.remove "foo")
```

> The goal being that it's non-trivial to work out the compiler errors when certain compositions don't compose correctly because of the laws and the compiler is forcing the developer to read function signatures for what look like dot-properties at the call site.
Just attempting to make things easier to reason about.

Not sure what the error messages look like. In terms of error messages, what would you want it to look like?
